File: kubwa-annotations/src/main/java/com/wesleyelliott/kubwa/annotation/Regex.java
package com.wesleyelliott.kubwa.annotation;

import com.wesleyelliott.kubwa.rule.RegexRule;

import java.lang.annotation.Documented;
import java.lang.annotation.Retention;
import java.lang.annotation.Target;

import static java.lang.annotation.ElementType.ANNOTATION_TYPE;
import static java.lang.annotation.ElementType.CONSTRUCTOR;
import static java.lang.annotation.ElementType.FIELD;
import static java.lang.annotation.ElementType.METHOD;
import static java.lang.annotation.ElementType.PARAMETER;
import static java.lang.annotation.ElementType.TYPE;
import static java.lang.annotation.RetentionPolicy.RUNTIME;

/**
 * Created by wesley on 2016/07/28.
 */
@ValidateUsing(RegexRule.class)
@Retention(RUNTIME)
@Target({ METHOD, FIELD, ANNOTATION_TYPE, CONSTRUCTOR, PARAMETER , TYPE})
public @interface Regex {
    int errorMessage();
    String name() default "regexError";
    String regex() default "*";

    @Target(TYPE)
    @Retention(RUNTIME)
    @Documented
    public @interface List {
        Regex[] value();
    }
}

File: kubwa-annotations/src/main/java/com/wesleyelliott/kubwa/annotation/Regex.kt
package com.wesleyelliott.kubwa.annotation

import com.wesleyelliott.kubwa.rule.RegexRule

/**
 * Created by wesley on 2016/07/28.
 */
@ValidateUsing(RegexRule::class)
@Retention(AnnotationRetention.RUNTIME)
@Target(
    AnnotationTarget.FUNCTION,
    AnnotationTarget.PROPERTY_GETTER,
    AnnotationTarget.PROPERTY_SETTER,
    AnnotationTarget.FIELD,
    AnnotationTarget.ANNOTATION_CLASS,
    AnnotationTarget.CONSTRUCTOR,
    AnnotationTarget.VALUE_PARAMETER,
    AnnotationTarget.CLASS,
    AnnotationTarget.FILE
)
annotation class Regex(
    val errorMessage: Int,
    val name: String = "regexError",
    val regex: String = "*"
) {

    @Target(AnnotationTarget.CLASS, AnnotationTarget.FILE)
    @Retention(AnnotationRetention.RUNTIME)
    @MustBeDocumented
    annotation class List(vararg val value: Regex)
}
